doc/ftpcount.1
.\"
.\" Copyright (c) 1999-2003 WU-FTPD Development Group. 
.\" All rights reserved.
.\" 
.\" Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994 
.\" The Regents of the University of California.  Portions Copyright (c) 
.\" 1993, 1994 Washington University in Saint Louis.  Portions Copyright 
.\" (c) 1996, 1998 Berkeley Software Design, Inc.  Portions Copyright (c) 
.\" 1998 Sendmail, Inc.  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric 
.\" P. Allman.  Portions Copyright (c) 1989 Massachusetts Institute of 
.\" Technology.  Portions Copyright (c) 1997 Stan Barber.  Portions 
.\" Copyright (C) 1991, 1992, 1993, 1994, 1995, 1996, 1997 Free Software 
.\" Foundation, Inc.  Portions Copyright (c) 1997 Kent Landfield. 
.\"
.\" Use and distribution of this software and its source code are governed 
.\" by the terms and conditions of the WU-FTPD Software License ("LICENSE"). 
.\"
.\"     $Id: ftpcount.1,v 1.11 2016/03/11 09:35:31 wmaton Exp $
.\"
.TH ftpcount 1
.SH NAME
ftpcount \- show current number of users for each class
.SH SYNTAX
.B ftpcount
[
.B \-v
] [
.B \-V
]
.SH DESCRIPTION
The
.B ftpcount
command
shows the current number of users (and the limit) for each
class defined in the ftpaccess file.
.PP
The
.B \-V
option causes the program to display copyright and version information, then
terminate.
.PP
If the
.B \-v
option is specified, the user counts for classes defined in virtual host
ftpaccess files are shown.
.SH SEE ALSO
ftpwho(1)


CHANGES
 
  Copyright (c) 1999-2003 WU-FTPD Development Group.  
  All rights reserved.
  
  Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994
    The Regents of the University of California.
  Portions Copyright (c) 1993, 1994 Washington University in Saint Louis.
  Portions Copyright (c) 1996, 1998 Berkeley Software Design, Inc.
  Portions Copyright (c) 1989 Massachusetts Institute of Technology.
  Portions Copyright (c) 1998 Sendmail, Inc.
  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric P.  Allman.
  Portions Copyright (c) 1997 Stan Barber.
  Portions Copyright (c) 1997 Kent Landfield.
  Portions Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997
    Free Software Foundation, Inc.  
 
  Use and distribution of this software and its source code are governed 
  by the terms and conditions of the WU-FTPD Software License ("LICENSE").
 
  If you did not receive a copy of the license, it may be obtained online
  at http://www.wu-ftpd.info/license.html.
 
  $Id: CHANGES,v 1.47 2016/03/14 23:57:54 wmaton Exp $

###########################################################################

Changes in 2.8.0-CC5: Released

 o  Updates to config.guess and config.sub to modernise the autoconf
    build system.

 o  configure script regenerated.

 o  Describe auto-buf tuning configure option in README.AUTOCONF.

 o  Add option (-R) to disable REST command in cases where download
    managers can abuse the server by opening multiple connections to
    retrieve chunks of the same file.  Based on a patch by Philippe Troin
    (See Debian bug ID 138497).

 o  Some minor man page cleanups.

 o  More syncs to Debian patches dating back to 2.4.2.

 o  Update INSTALL instructions.

 o  Changed the TLS default cipher from ALL:!EXP to exclude most insecure
    or unsupported ciphers within modern versions of OpenSSL.

Changes in 2.8.0-CC4: Released 7 January, 2012

 o  Thanks to a patch by Laurent CONSTANTIN, calling SSL_library_init()
    will no longer create an error within the SSL code related to a lack
    of ciphers.  The patch is applied to src/tlsutil.c.

 o  Another patch by Laurent CONSTANTIN, this time eliminating extra
    carriage returns that aren't necessary when calling syslog().

 o  Added -c option which displays the compile-time options that were used
    to build WU-FTPD. 

Changes in 2.8.0-CC3: Released 6 November, 2011

 o  Many URL updates.

 o  The default email address used by the server in absence of an email
    directive in ftpaccess is ftp-bugs, and is noted in ftpaccess now.

 o  Added code for ADLER-32 checksums for future extension to the list
    of algorithms available to do checksums.

 o  hard-coded links to cksum and md5sum within the FTP root now point to
    these utilities in /bin instead of /usr/bin.

 o  As defined in GFD.20 (GridFTP:  Protocol Extension to FTP for the
    Grid) by the Global Grid Forum, the SBUF command has been introduced.
    This calls the same code as SITE BUFSIZE would.  SITE BUFSIZE is
    still present for backwards compatibility.

 o  Thanks to a contribution by Austin Ellis, added RHEL6 packaging info
    to the codebase in packages/rpm.  This also includes a patch for
    building on that particular system.

 o  Updated ftpd.8 manpage documenting some missing SITE commands that
    had been in the server since 1998.

 o  First cut at implementing the remainder of the RFC3659 extended FTP
    commands.  FEAT, MLST, OPTS and MLSD are now supported.  Based upon
    early gridFTP releases which used WU-FTPD 2.6.0 and 2.6.1 code.

 o  According to draft-somers-ftp-mfxx-04.txt, defined the stubs for the
    MFF, MFCT and MFMT commands.  These may or may not be supported in the
    future, but at the least they are defined for now.  If so, they may
    become SITE commands instead if the spec is not approved.

 o  Updated doc/examples/ftpconversions based on Debian changes introduced
    sometime in 1995.  The server can now deal with such things as bzip2,
    xz, zip and unzip.

 o  If there's an empty /etc/fstab, a segfault can be produced.  A fix
    from Debian BugID 568235 to path_to_device in src/extensions.c addresses
    this.  Patch by Chris Butler.

 o  Added another possible path for quota.h, patch by Chris Butler of
    Debian.

 o  Added missing handler for SIGTERM, patch by Chris Butler of Debian
    related to Debian bug #536774.

 o  Changed -h invocation on the commandline to print out a very brief help
    listing all commandline options that could be supplied to the daemon.
    Also note that some switches are or have an effect on the configuration
    file.  The former function assigned to the 'h' option is now 'H'.

 o  The former function assigned to the 'h' option is now 'H'.  Said it
    twice now.  The manpage has been updated to reflect the changes.

 o  Implemented the CSID command as found in the IETF draft
    draft-peterson-streamlined-ftp-command-extensions.  This is meant
    to be a standards-based replacement for the CLNT command.  Currently
    this command is hard-wired with the server name and version, but the 
    next release will permit customisation.

###########################################################################

Changes in 2.8.0-CC2: Released 26 June, 2009

 o  Cleaned-up and updated the ftpd.8 manpage which had languished for
    some time.

 o  Added Auto tuning/auto buf, with added commands SITE BUFSIZE and
    SITE BUFSIZEMEASURE.  Use --enable-autobuf to enable this feature.
    Based on a patch on work done on WU-FTPD-2.6.1 by Gaurav Navlakha, et al
    of DAST, NLANR (now CAIDA).  This code also contains the ability to
    report back on buffer usage.  Note that an autobuf client needs to
    be used to take advantage of this feature.  For the time being, the
    code for buffer measurement is IPv4 only.

 o  Updated src/extensions.c to take into account a change in Linux quota
    structure.  Based on a patch by Chris Butler.


src/ftpcount.c
/****************************************************************************  
 
  Copyright (c) 1999-2003 WU-FTPD Development Group.  
  All rights reserved.
  
  Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994
    The Regents of the University of California.
  Portions Copyright (c) 1993, 1994 Washington University in Saint Louis.
  Portions Copyright (c) 1996, 1998 Berkeley Software Design, Inc.
  Portions Copyright (c) 1989 Massachusetts Institute of Technology.
  Portions Copyright (c) 1998 Sendmail, Inc.
  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric P.  Allman.
  Portions Copyright (c) 1997 by Stan Barber.
  Portions Copyright (c) 1997 by Kent Landfield.
  Portions Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997
    Free Software Foundation, Inc.  
 
  Use and distribution of this software and its source code are governed 
  by the terms and conditions of the WU-FTPD Software License ("LICENSE").
 
  If you did not receive a copy of the license, it may be obtained online
  at http://www.wu-ftpd.info/license.html.
 
  $Id: ftpcount.c,v 1.11 2011/11/28 01:54:15 wmaton Exp $
 
****************************************************************************/
#include "config.h"

#if defined(TSOL)
#  include <tsol/priv.h>
#endif /* defined(TSOL) */ 

#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#if defined(HAVE_SYS_SYSLOG_H)
#  include <sys/syslog.h>
#endif /* defined(HAVE_SYS_SYSLOG_H) */ 
#if defined(HAVE_SYSLOG_H) || (!defined(AUTOCONF) && !defined(HAVE_SYS_SYSLOG_H))
#  include <syslog.h>
#endif /* defined(HAVE_SYSLOG_H) || (!defined(AUTOCONF) && !defined(HAVE_SYS_SYSLOG_H)) */ 
#include <signal.h>
#include <time.h>
#include <ctype.h>
#include <limits.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <sys/param.h>

#if defined(HAVE_PATHS_H)
#  include <paths.h>
#endif /* defined(HAVE_PATHS_H) */ 

#if defined(VIRTUAL) && defined(INET6)
#  include <netinet/in.h>
#endif /* defined(VIRTUAL) && defined(INET6) */ 

#include "pathnames.h"
#include "extensions.h"

#if defined(HAVE_FCNTL_H)
#  include <fcntl.h>
#endif /* defined(HAVE_FCNTL_H) */ 

#if !defined(HAVE_STRLCPY) || !defined(HAVE_STRLCPY_PROTOTYPE)
size_t strlcpy(char *dst, const char *src, size_t siz);
#endif /*!defined(HAVE_STRLCPY) || !defined(HAVE_STRLCPY_PROTOTYPE)*/ 

#if defined(VIRTUAL)
extern int read_servers_line(FILE *, char *, size_t, char *, size_t);
#endif /* defined(VIRTUAL) */ 

#if defined(VIRTUAL)
#define ARGS	"Vv"
#else /* !(defined(VIRTUAL)) */ 
#define ARGS	"V"
#endif /* !(defined(VIRTUAL)) */ 

struct c_list {
    char *class;
    struct c_list *next;
};

void print_copyright(void);

char *progname;

/*************************************************************************/
/* FUNCTION  : parse_time                                                */
/* PURPOSE   : Check a single valid-time-string against the current time */
/*             and return whether or not a match occurs.                 */
/* ARGUMENTS : a pointer to the time-string                              */
/*************************************************************************/

static int parsetime(char *whattime)
{
    static char *days[] =
    {"Su", "Mo", "Tu", "We", "Th", "Fr", "Sa", "Wk"};
    time_t clock;
    struct tm *curtime;
    int wday, start, stop, ltime, validday, loop, match;

    (void) time(&clock);
    curtime = localtime(&clock);
    wday = curtime->tm_wday;
    validday = 0;
    match = 1;

    while (match && isalpha(*whattime) && isupper(*whattime)) {
	match = 0;
	for (loop = 0; loop < 8; loop++) {
	    if (strncmp(days[loop], whattime, 2) == 0) {
		whattime += 2;
		match = 1;
		if ((wday == loop) || ((loop == 7) && wday && (wday < 6)))
		    validday = 1;
	    }
	}
    }

    if (!validday) {
	if (strncmp(whattime, "Any", 3) == 0) {
	    validday = 1;
	    whattime += 3;
	}
	else
	    return (0);
    }

    if (sscanf(whattime, "%d-%d", &start, &stop) == 2) {
	ltime = curtime->tm_min + 100 * curtime->tm_hour;
	if ((start < stop) && ((ltime >= start) && ltime < stop))
	    return (1);
	if ((start > stop) && ((ltime >= start) || ltime < stop))
	    return (1);
    }
    else
	return (1);

    return (0);
}

/*************************************************************************/
/* FUNCTION  : validtime                                                 */
/* PURPOSE   : Break apart a set of valid time-strings and pass them to  */
/*             parse_time, returning whether or not ANY matches occurred */
/* ARGUMENTS : a pointer to the time-string                              */
/*************************************************************************/

static int validtime(char *ptr)
{
    char *nextptr;
    int good;

    while (1) {
	nextptr = strchr(ptr, '|');
	if (strchr(ptr, '|') == NULL)
	    return (parsetime(ptr));
	*nextptr = '\0';
	good = parsetime(ptr);
	*nextptr++ = '|';	/* gotta restore the | or things get skipped! */
	if (good)
	    return (1);
	ptr = nextptr;
    }
}

static int acl_fgetlimit(char *aclbuf, char *class)
{
    char *crptr, *ptr, linebuf[1024];
    int limit;

    while (*aclbuf != '\0') {
	if (strncasecmp(aclbuf, "limit", 5) == 0) {
	    for (crptr = aclbuf; *crptr++ != '\n';);
	    *--crptr = '\0';
	    strlcpy(linebuf, aclbuf, sizeof(linebuf));
	    *crptr = '\n';
	    (void) strtok(linebuf, " \t");	/* returns "limit" */
	    if ((ptr = strtok(NULL, " \t")) && (strcmp(class, ptr) == 0)) {
		if ((ptr = strtok(NULL, " \t"))) {
		    limit = atoi(ptr);	/* returns limit <n> */
		    if ((ptr = strtok(NULL, " \t")) && validtime(ptr))
			return (limit);
		}
	    }
	}
	while (*aclbuf && *aclbuf++ != '\n');
    }

    return (-1);
}

/*************************************************************************/
/* FUNCTION  : lock_fd                                                   */
/* PURPOSE   : Lock a file.                                              */
/* ARGUMENTS : File descriptor of file to lock.                          */
/*************************************************************************/

static void lock_fd(int fd)
{
#if !defined(HAVE_FLOCK)
    struct flock arg;
#endif /* !defined(HAVE_FLOCK) */ 

#if defined(HAVE_FLOCK)
    while (flock(fd, LOCK_SH)) {
#  if !defined(NO_PID_SLEEP_MSGS)
	syslog(LOG_ERR, "sleeping: flock of pid file failed: %m");
#  endif /* !defined(NO_PID_SLEEP_MSGS) */ 
#else /* !(defined(HAVE_FLOCK)) */ 
    arg.l_type = F_RDLCK;
    arg.l_whence = arg.l_start = arg.l_len = 0;
#  if defined(TSOL)
    if (set_effective_priv(PRIV_ON, 1, PRIV_FILE_LOCK) != 0) {
	syslog(LOG_ERR, "Cannot add PRIV_FILE_LOCK to eff. priv. set");
    }
#  endif /* defined(TSOL) */ 
    while (-1 == fcntl(fd, F_SETLK, &arg)) {
#  if !defined(NO_PID_SLEEP_MSGS)
	syslog(LOG_ERR, "sleeping: fcntl lock of pid file failed: %m");
#  endif /* !defined(NO_PID_SLEEP_MSGS) */ 
#endif /* !(defined(HAVE_FLOCK)) */ 
	sleep(1);
    }
#if !defined(HAVE_FLOCK)
#  if defined(TSOL)
    if (set_effective_priv(PRIV_OFF, 1, PRIV_FILE_LOCK) != 0) {
	syslog(LOG_ERR, "Cannot remove PRIV_FILE_LOCK from eff. priv. set");
    }
#  endif /* defined(TSOL) */ 
#endif /* !defined(HAVE_FLOCK) */ 
}

/*************************************************************************/
/* FUNCTION  : unlock_fd                                                 */
/* PURPOSE   : Unlock a file locked by lock_fd.                          */
/* ARGUMENTS : File descriptor of file to unlock.                        */
/*************************************************************************/

static void unlock_fd(int fd)
{
#if !defined(HAVE_FLOCK)
    struct flock arg;
#endif /* !defined(HAVE_FLOCK) */ 

#if defined(HAVE_FLOCK)
    flock(fd, LOCK_UN);
#else /* !(defined(HAVE_FLOCK)) */ 
    arg.l_type = F_UNLCK;
    arg.l_whence = arg.l_start = arg.l_len = 0;
#  if defined(TSOL)
    if (set_effective_priv(PRIV_ON, 1, PRIV_FILE_LOCK) != 0) {
	syslog(LOG_ERR, "Cannot add PRIV_FILE_LOCK to eff. priv. set");
    }
#  endif /* defined(TSOL) */ 
    fcntl(fd, F_SETLK, &arg);
#  if defined(TSOL)
    if (set_effective_priv(PRIV_OFF, 1, PRIV_FILE_LOCK) != 0) {
	syslog(LOG_ERR, "Cannot remove PRIV_FILE_LOCK from eff. priv. set");
    }
#  endif /* defined(TSOL) */ 
#endif /* !(defined(HAVE_FLOCK)) */ 
}

static int acl_countusers(char *class)
{
    int i, j, n, count, pidfd;
    pid_t procid;
    char pidfile[MAXPATHLEN];
    char line[1024];
    FILE *ZeFile;
    struct pidfile_header hdr;
    struct stat pinfo;
    unsigned char bits, *buf;

    snprintf(pidfile, sizeof(pidfile), _PATH_PIDNAMES, class);
    pidfd = open(pidfile, O_RDONLY);
    if (pidfd == -1) {
	return (0);
    }

    lock_fd(pidfd);
    if (read(pidfd, (void *)&hdr, sizeof(hdr)) != sizeof(hdr)) {
	unlock_fd(pidfd);
	close(pidfd);
	return (0);
    }
    if (strcmp(progname, "ftpcount") == 0) {
	unlock_fd(pidfd);
	close(pidfd);
	return (hdr.count);
    }

    /*
     * Printing the process information can take a long time, and while we
     * hold the lock no users can join or leave this class. To minimize the
     * problem, read the whole PID file into memory then release the lock.
     */
    if (fstat(pidfd, &pinfo) != 0) {
	unlock_fd(pidfd);
	close(pidfd);
        return (0);
    }
    if ((buf = malloc((size_t)pinfo.st_size)) == NULL) {
	unlock_fd(pidfd);
	close(pidfd);
        return (0);
    }
    n = read(pidfd, buf, (size_t)pinfo.st_size);
    unlock_fd(pidfd);
    close(pidfd);
    count = 0;
    procid = 0;
    for (i = 0; i < n; i++) {
	if (buf[i] == 0) {
	    procid += CHAR_BIT;
	}
	else {
	    bits = 1;
	    for (j = 0; j < CHAR_BIT; j++) {
		if (((buf[i] & bits) != 0) &&
		    ((kill(procid, 0) == 0) || (errno == EPERM))) {
#if defined(SVR4)
#  if defined(AIX)
		    snprintf(line, sizeof(line), "/bin/ps %d", procid);
#    elif defined(sun)
		    snprintf(line, sizeof(line), "/usr/ucb/ps auxww %ld", procid);
#  else /* !(defined(AIX)) */ 
#    if defined (LINUX)
		    snprintf(line, sizeof(line), "/bin/ps fwwwp %d", procid);
#    else /* !(defined (LINUX) */ 
		    snprintf(line, sizeof(line), "/bin/ps -f -p %d", procid);
#    endif /* !(defined (LINUX) */ 
#  endif /* !(defined(AIX)) */ 
#  elif defined(M_UNIX)
		    snprintf(line, sizeof(line), "/bin/ps -f -p %d", procid);
#else /* !(defined(SVR4)) */ 
		    snprintf(line, sizeof(line), "/bin/ps %d", procid);
#endif /* !(defined(SVR4)) */ 
		    ZeFile = popen(line, "r");
		    (void) fgets(line, sizeof(line), ZeFile);
		    line[0] = '\0';
		    (void) fgets(line, sizeof(line), ZeFile);
		    if (line[0] != '\0') {
			size_t i;
			for (i = strlen(line); (i > 0) && ((line[i - 1] == ' ') || (line[i - 1] == '\n')); --i)
			    line[i - 1] = '\0';
			printf("%s\n", line);
			count++;
		    }
		    pclose(ZeFile);
		}
		bits <<= 1;
		procid++;
	    }
	}
    }
    free(buf);
    return (count);
}

static void new_list(struct c_list **list)
{
    struct c_list *cp, *tcp;

    if (*list == NULL) {
	*list = (struct c_list *) malloc(sizeof(struct c_list));
	if (*list == NULL) {
	    perror("malloc error in new_list");
	    exit(1);
	}
    }
    else {
	cp = (*list)->next;
	while (cp) {
	    if (cp->class)
		free(cp->class);
	    tcp = cp;
	    cp = cp->next;
	    free(tcp);
	}
    }
    (*list)->next = NULL;
}

static int add_list(char *class, struct c_list **list)
{
    struct c_list *cp;

    for (cp = (*list)->next; cp; cp = cp->next) {
	if (!strcmp(cp->class, class))
	    return (-1);
    }

    cp = (struct c_list *) malloc(sizeof(struct c_list));
    if (cp == NULL) {
	perror("malloc error in add_list");
	exit(1);
    }

    cp->class = strdup(class);
    if (cp->class == NULL) {
	perror("malloc error in add_list");
	exit(1);
    }
    cp->next = (*list)->next;
    (*list)->next = cp;
    return (1);
}

static int display_info(char *ftpaccess, char *address)
{
    FILE *accessfile;
    char class[80], linebuf[1024], *aclbuf, *myaclbuf, *crptr;
    int limit;
    struct stat finfo;
    static struct c_list *list = NULL;

    if ((accessfile = fopen(ftpaccess, "r")) == NULL) {
	if (errno != ENOENT)
	    fprintf(stderr, "%s: could not open access file %s: %s\n",
		    progname, ftpaccess, strerror(errno));
	return (1);
    }
    if (fstat(fileno(accessfile), &finfo) != 0) {
	fprintf(stderr, "%s: could not fstat() access file %s: %s\n",
		progname, ftpaccess, strerror(errno));
	fclose(accessfile);
	return (1);
    }

    if (finfo.st_size == 0) {
	printf("%s: no service classes defined, no usage count kept\n", progname);
	fclose(accessfile);
	return (0);
    }
    else {
	if (!(aclbuf = (char *) malloc((size_t) finfo.st_size + 1))) {
	    fprintf(stderr, "%s: could not malloc aclbuf: %s\n",
		    progname, strerror(errno));
	    fclose(accessfile);
	    return (1);
	}
	fread(aclbuf, (size_t) finfo.st_size, 1, accessfile);
	fclose(accessfile);
	*(aclbuf + (size_t) finfo.st_size) = '\0';
    }

    (void) new_list(&list);
    myaclbuf = aclbuf;
    while (*myaclbuf != '\0') {
	if (strncasecmp(myaclbuf, "class", 5) == 0) {
	    for (crptr = myaclbuf; *crptr++ != '\n';);
	    *--crptr = '\0';
	    strlcpy(linebuf, myaclbuf, sizeof(linebuf));
	    *crptr = '\n';
	    (void) strtok(linebuf, " \t");	/* returns "class" */
	    /* returns class name */
	    strlcpy(class, strtok(NULL, " \t"), sizeof(class));
	    if ((add_list(class, &list)) < 0) {
		/* we have a class with multiple "class..." lines so, only
		 * display one count... */
		;
	    }
	    else {
		limit = acl_fgetlimit(myaclbuf, class);
#if defined(VIRTUAL)
		if (address != NULL)
		    printf("%s ", address);
#endif /* defined(VIRTUAL) */ 
		if (strcmp(progname, "ftpcount")) {
		    printf("Service class %s: \n", class);
		    printf("   - %3d users ", acl_countusers(class));
		}
		else {
		    printf("Service class %-20.20s - %3d users ",
			   class, acl_countusers(class));
		}
		if (limit == -1)
		    printf("(no maximum)\n");
		else
		    printf("(%3d maximum)\n", limit);
	    }
	}
	while (*myaclbuf && *myaclbuf++ != '\n');
    }
    free(aclbuf);
    return (0);
}

int main(int argc, char **argv)
{
    int c, exitval;
    int virtual = 0;
#if defined(VIRTUAL)
    FILE *svrfp;
    char *sp;
    struct stat st;
    char configdir[MAXPATHLEN];
    char accesspath[MAXPATHLEN];
#  if defined(INET6)
    char hostaddress[INET6_ADDRSTRLEN];
#  else /* !(defined(INET6)) */ 
    char hostaddress[32];
#  endif /* !(defined(INET6)) */ 
#endif /* defined(VIRTUAL) */ 

#if defined(TSOL)
/* Before anything, clear the effective privilege set */

    if (set_effective_priv(PRIV_SET, 0) != 0) {
	syslog(LOG_ERR, "ftp[count|who] cannot clear effective privileges!");
	exit(1);
    }
#endif /* defined(TSOL) */ 

    if ((progname = strrchr(argv[0], '/')))
	++progname;
    else
	progname = argv[0];

    if (argc > 1) {
	while ((c = getopt(argc, argv, ARGS)) != EOF) {
	    switch (c) {
	    case 'V':
		print_copyright();
		exit(0);
#if defined(VIRTUAL)
	    case 'v':
		virtual = 1;
		break;
#endif /* defined(VIRTUAL) */ 
	    default:
		fprintf(stderr, "usage: %s [-" ARGS "]\n", progname);
		exit(1);
	    }
	}
    }

    exitval = 0;
    if ((virtual == 0) && (display_info(_PATH_FTPACCESS, NULL) != 0))
	exitval = 1;

#if defined(VIRTUAL)
    /*
     * Deal with the ftpaccess files at the virtual domain directory locations
     * specified in the ftpservers file.
     */
    if (virtual && ((svrfp = fopen(_PATH_FTPSERVERS, "r")) != NULL)) {
	while (read_servers_line(svrfp, hostaddress, sizeof(hostaddress), configdir, sizeof(configdir)) == 1) {
	    /* get rid of any trailing slash */
	    sp = configdir + (strlen(configdir) - 1);
	    if (*sp == '/')
		*sp = '\0';

	    /* check to see that a valid directory value was supplied */
	    if ((stat(configdir, &st) == 0) &&
		((st.st_mode & S_IFMT) == S_IFDIR)) {
		snprintf(accesspath, sizeof(accesspath), "%s/ftpaccess",
			 configdir);
		if (display_info(accesspath, hostaddress) != 0)
		    exitval = 1;
	    }
	}
	fclose(svrfp);
    }
#endif /* defined(VIRTUAL) */ 
    return (exitval);
}
